### PR TITLE
Workload generator, measurement infrastructure and utilization plotting

### PR DESCRIPTION
**Pull request by LedgeDash**
## workload generator (updated)
Generate workload by specifying
1. average inter-arrival time (`mu`) in ms 
2. spike window (expressed in `start time` and `end time`)

for each function in a workload characteristic YAML file. 

Call it with: `python3 generator.py <workload.yaml> <request.json>`
`<workload.yaml>` is the input.
`<request.json>` is the output request file

Here's an example `example_workload.yaml`：
```yaml
- name: "loremjs"
  mu: 2 # average inter-arrival time in ms. 1000/mu=arrival rate per sec
  start_time: 0 # ms
  end_time: 5000 # ms                                                   
- name: "lorempy2"
  mu: 2
  start_time: 5000
  end_time: 10000
```
This YAML generates a workload of 2 functions `loremjs` and `lorempy2`. `loremjs`'s spike begin at timestamp=0ms and ends at timestamp=5000ms. `lorempy2`'s spike begin at timestamp=5000ms and ends at timestamp=10000ms. During non-spike periods (that is [5000, 10000] for `loremjs` and [0,5000] for `lorempy2`), functions have a default average inter-arrival time of 1000ms (i.e., average of 1 req/sec), currently hard-coded. 

If you want a function to have multiple spikes, you would for example do this:
```yaml
- name: "loremjs"
  mu: 2 # average inter-arrival time in ms. 1000/mu=arrival rate per sec
  start_time: 0 # ms
  end_time: 5000 # ms                                                   
- name: "lorempy2"
  mu: 2
  start_time: 5000
  end_time: 10000
- name: "loremjs"
  mu: 2 
  start_time: 7000 # ms
  end_time: 15000 # ms 
```
It is desirable to list functions in chronological order so that the workload YAML file basically outlines the timeline of the workload.

## Time measurement
In order to measure utilization over time (think of a plot of utilization where x-axis is time), I added code in controller that outputs timestamps when certain events happen. These events are: VM boot start (calling the `.run()` function on `VmAppConfig`), controller receiving the tty ready message from a VM, request sent, response received, VM eviction start, VM eviction finished.

All results are output as a json string to `./measurement/measurement-<start_time>-<end_time>.json` file, where <start_time> is the timestamp of experiment start time (i.e., right before the first request is scheduled) and <end_time> is the timestamp of experiment finished.

In the json string, you will see something like this:
```json
{
  "boot timestamps": {
    "10": [
      14283896112730,
      14284027647863
    ],
    "11": [
      14283907501412,
      14283983962431
    ],
  ...
 },
"eviction timestamps": {
  "7": [
    14283953935355,
    14283991563487
  ]
},
"request/response timestamps": {
  "10": [
    14283897402698,
    14284037372778,
    14284042663022,
    14284045126842,
    14284052737579,
    14284055072850
  ],
...
}

}
```
(the `...` just means there are more elements, but omitted to save space)
For example, `"request/response timestamps"` object has all the timestamp of all requests and response from and to all VMs. The `"10"` is VM id. the 6 numbers are 3 pairs of (request_send_timestamp, response_received_timestamp), sequentially. 

With all these information, we then know exactly what happened when throughput the entire duration of a workload. The `plot.py` script takes this information to calculate and plot utilization.

Next steps (need Cgroup CPU share feature completed first):
1. plot.py need to also take in function config yaml file to know each function's memory requirement
2. utilization calculation need to base off of total memory size, not core count

### CODE REVIEW

**Comment by alevy** [comment]
A lot of the changes in this PR are now unrelated to the title so it's hard to review. I have some substantive comments about the workload generator, and I will mark up the rest as I see things, but I think we should merge this ASAP

**Comment by LedgeDash** [comment]
Yes, I think it'd be great to merge this asap.
I've updated the title since the PR content has changed quite significantly. Nevertheless, the workload generator code is separate enough from measurement and plotting related code such that we could just treat this PR as 2 PRs combined. 

I'm updating the PR description (the first comment) to include my new changes.

File: workloads/generator.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import numpy as np
```
Comment (alevy) [inline workloads/generator.py:1]: This script is a bit difficult to follow, so I'm annotating in comments what I think is going on.

File: workloads/generator.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+print('function names: ' + str(function_names))
+print('arrival rates: ' + str(arrival_rates))
+print('number of invocations: ' + str(num_invocations))
+
+
+mu = 1/arrival_rates
+num_functions = len(arrival_rates)
+num_invocations = num_invocations[0]
+inter_arrival_time = np.random.exponential(mu, (num_invocations, num_functions))
```
Comment (alevy) [inline workloads/generator.py]: We end up with a vector of n-tuples, each element of the vector has an element for each function that represents the the time delta between the last request of that function and "this" request.

File: workloads/generator.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+function_names = [f['name'] for f in data]
+arrival_rates = np.array([f['arrival_rate'] for f in data]) # num of invocations per second
+num_invocations = np.array([f['num_invocation'] for f in data])
+
+print('function names: ' + str(function_names))
+print('arrival rates: ' + str(arrival_rates))
+print('number of invocations: ' + str(num_invocations))
+
+
+mu = 1/arrival_rates
```
Comment (alevy) [inline workloads/generator.py]: `mu` is the scale parameter. Do we really want an inverse of the average arrival rate? in an exponential distribution, the average is `1 / lambda`. `np.random.exponent`'s first parameter (called `scale`) is also `1 / lambda`, so would we want it to be equal the average exactly? in other words, not take the inverse?

File: workloads/generator.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+num_functions = len(arrival_rates)
+num_invocations = num_invocations[0]
+inter_arrival_time = np.random.exponential(mu, (num_invocations, num_functions))
+inter_arrival_time = inter_arrival_time * 1000 # convert to ms
```
Comment (alevy) [inline workloads/generator.py]: Covert to milliseconds. Why do we need to do this? Are average arrival rates already expressed as milliseconds in the yaml file?

File: workloads/generator.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+inter_arrival_time = np.ceil(inter_arrival_time)
+inter_arrival_time_cumsum = np.cumsum(inter_arrival_time, axis=0)
```
Comment (alevy) [inline workloads/generator.py]: `cumsum` coverts the vector into a cumulative sum, meaning now each element represents the delta time from the beginning of the experiment. Basically a timestamp. Great.

At this point, what I would imaging doing is:

  a. Associating a function invocation with each element
  b. Flattening out the vector
  c. Sorting by timestamp
  d. Done! (outputting timestamp instead of delta from previous request requires changing the controller code, but that's fine)

I'm moving forward with this as my mental model for the order of elements we want in the final output.

File: workloads/generator.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+inter_arrival_time_cumsum = inter_arrival_time_cumsum.astype(int)
+#print(inter_arrival_time_cumsum )
+
+fd = open(output_request_file, 'w')
+
+pmin = 0
+
+while not finished(search_index, num_invocations):
+    candidates = [inter_arrival_time_cumsum[search_index[i], i] for i in range(num_functions)]
```
Comment (alevy) [inline workloads/generator.py]: in each iteration, this generates a tuple with one element for each function. Each element copies the element (a timestamp) from the arrival-time tuple indexed by the value of `search_index[i]`.

So `search_index` needs to keep track, for each function, which is the next element in inter_arrival_time_cumsum that hasn't been used for that particular function...

File: workloads/generator.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+search_index = np.zeros(num_functions, dtype=np.int32)
+
+inter_arrival_time_cumsum = inter_arrival_time_cumsum.astype(int)
+#print(inter_arrival_time_cumsum )
+
+fd = open(output_request_file, 'w')
+
+pmin = 0
+
+while not finished(search_index, num_invocations):
```
Comment (alevy) [inline workloads/generator.py]: Each iteration through the loop is going to output _one_ request for _one_ function.

File: workloads/generator.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    candidates = [inter_arrival_time_cumsum[search_index[i], i] for i in range(num_functions)]
+    minv = np.min(candidates)
```
Comment (alevy) [inline workloads/generator.py]: Given the current tuple of timestamps, choose the smallest one

File: workloads/generator.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    min_idx = np.argmin(candidates)
```
Comment (alevy) [inline workloads/generator.py]: min_idx is basically the function index for the minimum timestamp we just chose

File: workloads/generator.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+    interval = minv - pmin
```
Comment (alevy) [inline workloads/generator.py]: the interval is the current timestamp minus the last timestamp we just output (or 0 on the first iteration)

File: workloads/generator.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    pmin = minv
+
+    json.dump({"interval": int(interval), "function": function_names[min_idx], "payload":{"request": 42}}, fd)
+    fd.write('\n')
+
+    search_index[min_idx] = search_index[min_idx] + 1
```
Comment (alevy) [inline workloads/generator.py]: Update _only_ the search_index for the function we just output.

File: workloads/generator.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+    if search_index[min_idx] == num_invocations:
```
Comment (alevy) [inline workloads/generator.py]: I think I get what this is doing, but I think we can also safely get away with `num_invocations` in general. What we're looking for is a steady state distribution of invocations for different functions. It doesn't make sense for a function to just _completely_ stop being invoked. Instead maybe we want to cap requests after a certain amount of total experiment time or something.

**Review by alevy — COMMENTED** [review]
OK, I think I understand how the workload generator works, and I think it's basically _OK_ except for the choice of the `lambda` parameter. The algorithm itself seems a bit over-complicated, so it's hard to follow, but I do believe it does what we want.

File: workloads/generator.py (hunk 4 above)
Comment (LedgeDash) [reply workloads/generator.py]: So currently `arrival_rates` is number of invocations per _second_ (see line 28). But you're right, we should just change in the yaml file to have the arrival_rates mean invocations per millisecond and get rid of this `*1000` line. That way our time unit is consistent. I'll make the change

File: workloads/generator.py (hunk 12 above)
Comment (LedgeDash) [reply workloads/generator.py]: Yes! Great point! That would help measurement and plotting! I'll make the change.

File: workloads/generator.py (hunk 4 above)
Comment (alevy) [reply workloads/generator.py]: Oooohhhhhh.... I see.

File: workloads/generator.py (hunk 3 above)
Comment (LedgeDash) [reply workloads/generator.py]: hmm... I might be wrong but the documentation (which is really poorly worded...) says:
>\beta is the scale parameter, which is the inverse of the rate parameter \lambda = 1/\beta.

which I believe means that the input to the function is the average inter-arrival time. If this is the case, we would want to take the inverse, i.e., `mu = 1/arrival_rates`.

Comment (LedgeDash) [reply workloads/generator.py]: OK, so I compared the probability density function in the [numpy.random.exponential documentation](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.exponential.html) with [Wikipedia, “Exponential distribution”](http://en.wikipedia.org/wiki/Exponential_distribution). They're using the same notations.
So `scale` is `1/lambda`, where `lambda` is the arrival rate. So I believe taking the inverse is necessary.

Comment (alevy) [reply workloads/generator.py]: > where lambda is the arrival rate

I don't believe that's true. The mean of the distribution is `1/ lambda`, and the mean of the distribution should equal the average (mean) arrival rate, no?

Comment (alevy) [reply workloads/generator.py]: Oh, I'm confused because I misinterpreted `arrival_rates` as in the adjacent comment thread---what it _means_ is the number of times per second (a rate) that requests arrive on average. What I _thought_ it meant was the average delta in milliseconds between subsequent requests.

Comment (LedgeDash) [reply workloads/generator.py]: yes. sorry I should have been more clear in my comments (or added comments to `generator.py` in general). My bad... But just to make sure we're on the same page, there are 2 parameters:
1. The rate parameter `\lambda`, which is the `arrival_rates` variable in the code. This is the **average number of requests per unit of time**
2. The scale parameter `\beta`, which is the `mu` variable in the code. This is the inverse of `\lambda`, and it represents the **average inter-arrival time between requests**. It is also the mean of the exponential distribution (thus I named the variable `mu` in the code).

**Comment by LedgeDash** [comment]
I've updated the workload generator incorporating @alevy 's comments. Inspired by @alevy 's insight that we could get rid of `num_invocations` altogether. I changed the interface to specifying a spike window for each function. Details are in my updated PR description (first comment). This I believe makes the generator more flexible and nicely makes the workload YAML file a chronological outline of a workload which is more intuitive and easier for our experiments' purposes.

I also got rid of `arrival rates` as that was a source of confusion and switched directly to `mu` the average inter-arrival _time_. And all time values are now in ms.

**Review by alevy — APPROVED** [review]
Approving lieu of a full re-review since we're short on time. But seems correct, and @LedgeDash's been testing.
